## Dataset Information: 

### 1.(global_active_power*1000/60 - sub_metering_1 - sub_metering_2 - sub_metering_3) represents the active energy consumed every minute (in watt hour) in the household by electrical equipment not measured in sub-meterings 1, 2 and 3.

### 2.The dataset contains some missing values in the measurements (nearly 1,25% of the rows). All calendar timestamps are present in the dataset but for some timestamps, the measurement values are missing: a missing value is represented by the absence of value between two consecutive semi-colon attribute separators. For instance, the dataset shows missing values on April 28, 2007.

## Attribute Information:

#### 1.date: Date in format dd/mm/yyyy
#### 2.time: time in format hh:mm:ss
#### 3.global_active_power: household global minute-averaged active power (in kilowatt)
#### 4.global_reactive_power: household global minute-averaged reactive power (in kilowatt)
#### 5.voltage: minute-averaged voltage (in volt)
#### 6.global_intensity: household global minute-averaged current intensity (in ampere)
#### 7.sub_metering_1: energy sub-metering No. 1 (in watt-hour of active energy). It corresponds to the kitchen, containing mainly a dishwasher, an oven and a microwave (hot plates are not electric but gas powered).
#### 8.sub_metering_2: energy sub-metering No. 2 (in watt-hour of active energy). It corresponds to the laundry room, containing a washing-machine, a tumble-drier, a refrigerator and a light.
#### 9.sub_metering_3: energy sub-metering No. 3 (in watt-hour of active energy). It corresponds to an electric water-heater and an air-conditioner.

### importing libraries

In [1]:
import pandas as ps
import numpy as np

### Data ingestion

In [128]:
df = ps.read_csv(r'C:\Users\dchakrab.COMPUTACENTER\Downloads\household_power_consumption\household_power_consumption.csv', sep = ';')

C:\Users\dchakrab.COMPUTACENTER\AppData\Local\Temp\ipykernel_18916\937424890.py:1: DtypeWarning: Columns (2,3,4,5,6,7) have mixed types. Specify dtype option on import or set low_memory=False.
  df = ps.read_csv(r'C:\Users\dchakrab.COMPUTACENTER\Downloads\household_power_consumption\household_power_consumption.csv', sep = ';')


In [129]:
df.head()

Date      Time Global_active_power Global_reactive_power  Voltage  \
0  16/12/2006  17:24:00               4.216                 0.418  234.840   
1  16/12/2006  17:25:00               5.360                 0.436  233.630   
2  16/12/2006  17:26:00               5.374                 0.498  233.290   
3  16/12/2006  17:27:00               5.388                 0.502  233.740   
4  16/12/2006  17:28:00               3.666                 0.528  235.680   

  Global_intensity Sub_metering_1 Sub_metering_2  Sub_metering_3  
0           18.400          0.000          1.000            17.0  
1           23.000          0.000          1.000            16.0  
2           23.000          0.000          2.000            17.0  
3           23.000          0.000          1.000            17.0  
4           15.800          0.000          1.000            17.0

### Checking duplicated value

In [130]:
df.duplicated().sum()

0

### There is no duplicated value in the dataset

In [131]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075259 entries, 0 to 2075258
Data columns (total 9 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Date                   object 
 1   Time                   object 
 2   Global_active_power    object 
 3   Global_reactive_power  object 
 4   Voltage                object 
 5   Global_intensity       object 
 6   Sub_metering_1         object 
 7   Sub_metering_2         object 
 8   Sub_metering_3         float64
dtypes: float64(1), object(8)
memory usage: 142.5+ MB


### As we have seen that in this dataset we have 20,75,259 rows so then we can take a sample of dataset from it and work on it.

In [132]:
df = df.sample(n=50000, random_state=200, ignore_index=True)

### checking for null values

In [133]:
df.isnull().sum()

Date                       0
Time                       0
Global_active_power        0
Global_reactive_power      0
Voltage                    0
Global_intensity           0
Sub_metering_1             0
Sub_metering_2             0
Sub_metering_3           632
dtype: int64

#### We can see that Sub_metering_3 has 632 missing values or null values and we need to handle those missing values.
#### We can fill the null values with the mean values.

In [8]:
df['Sub_metering_3'].fillna(value = df['Sub_metering_3'].mean(), inplace = True)

In [13]:
## checking for null values now

In [9]:
df.isnull().sum()

Date                     0
Time                     0
Global_active_power      0
Global_reactive_power    0
Voltage                  0
Global_intensity         0
Sub_metering_1           0
Sub_metering_2           0
Sub_metering_3           0
dtype: int64

In [143]:
### There is some impurate value. We replace it with '0' so that it will not disturb 
df.replace(['?', np.nan], '0', inplace=True)

In [11]:
df.head()

Date      Time Global_active_power Global_reactive_power  Voltage  \
0   27/7/2007  10:25:00               0.196                 0.074  234.170   
1   28/8/2010  18:07:00               1.262                 0.054  241.420   
2   30/9/2008  00:48:00               0.344                 0.076  242.460   
3  12/10/2009  06:54:00               2.672                 0.000  238.840   
4   10/1/2010  23:56:00               0.254                 0.000  250.220   

  Global_intensity Sub_metering_1 Sub_metering_2  Sub_metering_3  
0            0.800          0.000          0.000             0.0  
1            5.200          0.000          0.000            18.0  
2            1.600          0.000          0.000             0.0  
3           11.400          0.000          0.000            18.0  
4            1.000          0.000          0.000             0.0

In [137]:
df['Date']=ps.to_datetime(df['Date'])

C:\Users\dchakrab.COMPUTACENTER\AppData\Local\Temp\ipykernel_18916\2729677711.py:1: UserWarning: Parsing '29/10/2009' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  df['Date']=ps.to_datetime(df['Date'])
C:\Users\dchakrab.COMPUTACENTER\AppData\Local\Temp\ipykernel_18916\2729677711.py:1: UserWarning: Parsing '13/11/2007' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  df['Date']=ps.to_datetime(df['Date'])
C:\Users\dchakrab.COMPUTACENTER\AppData\Local\Temp\ipykernel_18916\2729677711.py:1: UserWarning: Parsing '17/10/2009' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  df['Date']=ps.to_datetime(df['Date'])
C:\Users\dchakrab.COMPUTACENTER\AppData\Local\Temp\ipykernel_18916\2729677711.py:1: UserWarning: Parsing '30/11/2007' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  df['Date']

C:\Users\dchakrab.COMPUTACENTER\AppData\Local\Temp\ipykernel_18916\2729677711.py:1: UserWarning: Parsing '30/10/2008' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  df['Date']=ps.to_datetime(df['Date'])
C:\Users\dchakrab.COMPUTACENTER\AppData\Local\Temp\ipykernel_18916\2729677711.py:1: UserWarning: Parsing '23/11/2008' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  df['Date']=ps.to_datetime(df['Date'])
C:\Users\dchakrab.COMPUTACENTER\AppData\Local\Temp\ipykernel_18916\2729677711.py:1: UserWarning: Parsing '31/12/2008' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  df['Date']=ps.to_datetime(df['Date'])
C:\Users\dchakrab.COMPUTACENTER\AppData\Local\Temp\ipykernel_18916\2729677711.py:1: UserWarning: Parsing '26/10/2010' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  df['Date']

C:\Users\dchakrab.COMPUTACENTER\AppData\Local\Temp\ipykernel_18916\2729677711.py:1: UserWarning: Parsing '21/10/2007' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  df['Date']=ps.to_datetime(df['Date'])
C:\Users\dchakrab.COMPUTACENTER\AppData\Local\Temp\ipykernel_18916\2729677711.py:1: UserWarning: Parsing '13/12/2008' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  df['Date']=ps.to_datetime(df['Date'])
C:\Users\dchakrab.COMPUTACENTER\AppData\Local\Temp\ipykernel_18916\2729677711.py:1: UserWarning: Parsing '22/10/2008' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  df['Date']=ps.to_datetime(df['Date'])
C:\Users\dchakrab.COMPUTACENTER\AppData\Local\Temp\ipykernel_18916\2729677711.py:1: UserWarning: Parsing '24/11/2008' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  df['Date']

In [138]:
df['Time']=ps.to_datetime(df['Time'])

In [139]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   Date                   50000 non-null  datetime64[ns]
 1   Time                   50000 non-null  datetime64[ns]
 2   Global_active_power    50000 non-null  object        
 3   Global_reactive_power  50000 non-null  object        
 4   Voltage                50000 non-null  object        
 5   Global_intensity       50000 non-null  object        
 6   Sub_metering_1         50000 non-null  object        
 7   Sub_metering_2         50000 non-null  object        
 8   Sub_metering_3         49368 non-null  float64       
dtypes: datetime64[ns](2), float64(1), object(6)
memory usage: 3.4+ MB


In [140]:
df.columns

Index(['Date', 'Time', 'Global_active_power', 'Global_reactive_power',
       'Voltage', 'Global_intensity', 'Sub_metering_1', 'Sub_metering_2',
       'Sub_metering_3'],
      dtype='object')

In [42]:
### We need to convert object datatypes to numerics

In [144]:
for fea in df.columns:
    if df[fea].dtype=='O':
        df[fea]=ps.to_numeric(df[fea])

In [145]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   Date                   50000 non-null  datetime64[ns]
 1   Time                   50000 non-null  datetime64[ns]
 2   Global_active_power    50000 non-null  float64       
 3   Global_reactive_power  50000 non-null  float64       
 4   Voltage                50000 non-null  float64       
 5   Global_intensity       50000 non-null  float64       
 6   Sub_metering_1         50000 non-null  float64       
 7   Sub_metering_2         50000 non-null  float64       
 8   Sub_metering_3         50000 non-null  float64       
dtypes: datetime64[ns](2), float64(7)
memory usage: 3.4 MB


In [146]:
df.describe().T

count        mean        std  min      25%      50%  \
Global_active_power    50000.0    1.075417   1.057777  0.0    0.304    0.572   
Global_reactive_power  50000.0    0.122689   0.114243  0.0    0.046    0.098   
Voltage                50000.0  237.799407  27.097288  0.0  238.910  240.980   
Global_intensity       50000.0    4.558800   4.444099  0.0    1.400    2.600   
Sub_metering_1         50000.0    1.086660   6.029129  0.0    0.000    0.000   
Sub_metering_2         50000.0    1.255560   5.762398  0.0    0.000    0.000   
Sub_metering_3         50000.0    6.343720   8.402056  0.0    0.000    1.000   

                           75%      max  
Global_active_power      1.516    9.134  
Global_reactive_power    0.194    1.082  
Voltage                242.860  252.870  
Global_intensity         6.400   39.400  
Sub_metering_1           0.000   80.000  
Sub_metering_2           1.000   77.000  
Sub_metering_3          17.000   31.000

#### Now lets find out the total power consumption or the target feature of the dataset

In [147]:
tf = df['Global_active_power']*1000/60
tp = df['Sub_metering_1']+df['Sub_metering_2']+df['Sub_metering_2']

In [148]:
df['total_power_consumption']=tf-tp

#### Now lets group the total submetering value of the dataset

In [149]:
df['Sub_metering']=df['Sub_metering_1']+df['Sub_metering_2']+df['Sub_metering_3']

#### Now Dropping the features sub_metering_1, sub_metering_2, sub_metering_3

In [150]:
df.drop(['Sub_metering_1','Sub_metering_2','Sub_metering_3'],inplace = True, axis = 1)

In [151]:
df.head()

Date                Time  Global_active_power  Global_reactive_power  \
0 2007-07-27 2022-11-16 10:25:00                0.196                  0.074   
1 2010-08-28 2022-11-16 18:07:00                1.262                  0.054   
2 2008-09-30 2022-11-16 00:48:00                0.344                  0.076   
3 2009-12-10 2022-11-16 06:54:00                2.672                  0.000   
4 2010-10-01 2022-11-16 23:56:00                0.254                  0.000   

   Voltage  Global_intensity  total_power_consumption  Sub_metering  
0   234.17               0.8                 3.266667           0.0  
1   241.42               5.2                21.033333          18.0  
2   242.46               1.6                 5.733333           0.0  
3   238.84              11.4                44.533333          18.0  
4   250.22               1.0                 4.233333           0.0

In [ ]:
## Now we need to sepearte dependent and independent variables

In [152]:
X=df[['Global_active_power','Global_reactive_power','Voltage','Global_intensity']]

In [153]:
y=df['total_power_consumption']

### Now splitting training and test data

In [154]:
from sklearn.model_selection import train_test_split,GridSearchCV

In [155]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.33,random_state=10)

#### Importing Bagging Regressor

In [156]:
from sklearn.ensemble import BaggingRegressor

In [157]:
bagging_reg = BaggingRegressor()

In [ ]:
#### If base_estimator is none then by default base_estimator will be decision tree regression

In [158]:
bagging_reg.fit(X_train,y_train)

BaggingRegressor()

In [159]:
bag_reg_pred = bagging_reg.predict(X_test)

In [160]:
from sklearn.metrics import r2_score

### R2 Score for Bagging Regressor

In [161]:
r2_score(y_test,bag_reg_pred)

0.5827815413154552

In [162]:
bg_r2=r2_score(y_test,bag_reg_pred)

In [ ]:
Adjusted R2 score

In [163]:
ajd_r2=1-(1-bg_r2)*(len(y_test)-1)/(len(y_test)-X_test.shape[1]-1)

In [164]:
ajd_r2

0.5826803667877354

#### Importing RandomForest Regressor

In [40]:
from sklearn.ensemble import RandomForestRegressor

In [165]:
Rf_reg = RandomForestRegressor()

In [166]:
Rf_reg.fit(X_train,y_train)

RandomForestRegressor()

In [167]:
Rf_reg_pred = Rf_reg.predict(X_test)

In [168]:
Rf_reg_pred

array([ 3.238     ,  0.        ,  2.332     , ..., 23.93033333,
        4.65066667, 33.49666667])

In [45]:
r2_score(y_test,Rf_reg_pred)

0.6120228988119982

#### HyperParameter Tunning

In [179]:
params={
    "n_estimators":[15,10,50],
    "max_samples": range(2,10,1),
    "max_features":range(2,10,3)
}

In [170]:
grid_search=GridSearchCV(estimator=bagging_reg,param_grid=params,cv=3,verbose=2,n_jobs=-1)

In [171]:
grid_search.fit(X_train,y_train)

Fitting 3 folds for each of 192 candidates, totalling 576 fits


C:\Users\dchakrab.COMPUTACENTER\Anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
360 fits failed out of a total of 576.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
360 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\dchakrab.COMPUTACENTER\Anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\dchakrab.COMPUTACENTER\Anaconda3\lib\site-packages\sklearn\ensemble\_bagging.py", line 269, in fit
    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)
  File "C:\Users\dchakrab.COMPUTACENTER\Anaconda3\lib\site-packages\sklearn\ensembl

GridSearchCV(cv=3, estimator=BaggingRegressor(), n_jobs=-1,
             param_grid={'max_features': range(2, 10),
                         'max_samples': range(2, 10),
                         'n_estimators': [15, 10, 50]},
             verbose=2)

In [172]:
grid_search.best_params_

{'max_features': 3, 'max_samples': 9, 'n_estimators': 50}

In [173]:
bag_model_with_best_params= BaggingRegressor(n_estimators= 50, max_features = 3, max_samples = 9)

In [174]:
bag_model_with_best_params.fit(X_train,y_train)

BaggingRegressor(max_features=3, max_samples=9, n_estimators=50)

In [175]:
bag_y_predict = bag_model_with_best_params.predict(X_test)

In [176]:
bag_y_predict

array([ 5.45      ,  4.84133333,  3.976     , ..., 23.69733333,
        5.63466667, 27.33733333])

In [177]:
bag_model_r2_score = r2_score(y_test,bag_y_predict)

#### Post Hyper Parameter Tunning Bag Model R2 Score

In [178]:
bag_model_r2_score

0.5861105872145973

#### Post Hyper Parameter Tunning  Bag Model Adjusted R2 score

In [180]:
adj_r2_score = 1-(1-bag_model_r2_score)*(len(y_test)-1)/(len(y_test)-X_test.shape[1]-1)

In [181]:
adj_r2_score

0.5860102199729398

### Extra Tree Regressor

In [62]:
from sklearn.ensemble import ExtraTreesRegressor

In [63]:
ext_reg= ExtraTreesRegressor()

In [64]:
ext_reg.fit(X_train, y_train)

ExtraTreesRegressor()

In [65]:
ext_reg.predict(X_test)

array([ 3.34166667,  0.        ,  2.401     , ..., 23.31933333,
        4.393     , 33.77866667])

In [66]:
ext_reg_pred = ext_reg.predict(X_test)

### R2 Score on extra tree regressor

In [69]:
ext_tree_r2_score=r2_score(ext_reg_pred,y_test)

In [70]:
ext_tree_r2_score

0.40551039138706557

### Adjusted R2 Score on Extra Tree Regressor

In [71]:
ext_tree_adj_r2_score=1-(1-ext_tree_r2_score)*(len(y_test)-1)/(len(y_test)-X_test.shape[1]-1)

In [72]:
ext_tree_adj_r2_score

0.40536622900849917

### Voting Regressor

In [182]:
from sklearn.ensemble import VotingRegressor

In [183]:
rf = RandomForestRegressor(n_estimators=350,random_state=3)
et = ExtraTreesRegressor(n_estimators=100,random_state=3)
vt_reg= VotingRegressor([('rf', rf),('et',et)])

In [184]:
vt_reg.fit(X_train,y_train)

VotingRegressor(estimators=[('rf',
                             RandomForestRegressor(n_estimators=350,
                                                   random_state=3)),
                            ('et', ExtraTreesRegressor(random_state=3))])

In [185]:
vt_reg.predict(X_test)

array([ 3.15071429,  0.        ,  2.39333333, ..., 23.76378571,
        4.54109524, 33.12483333])

In [186]:
vt_pred = vt_reg.predict(X_test)

#### R2 Score

In [187]:
vt_r2_score = r2_score(vt_pred,y_test)

In [188]:
vt_r2_score

0.41150925325029897

#### Adjusted R2 Score

In [189]:
adj_vt_r2_score = 1-(1-vt_r2_score)*(len(y_test)-1)/(len(y_test)-X_test.shape[1]-1)

In [190]:
adj_vt_r2_score

0.4113665455820965